### ISSUE TITLE
Binaryen as a qemu backend

### ISSUE THREAD

**Issue by tbodt** [WebAssembly/binaryen#1494]
(Continuing conversation from WebAssembly/design#796)

I compiled qemu with `--enable-profiler`, added a few missing profiler logs, set an uninitialized variable to zero, and tested booting a FreeDOS image with 1 cpu (so no multithreading.) Booting to the C:\> prompt took 10 seconds. `info jit` at the monitor prompt gives this:

```
Translation buffer state:
gen code size       31204435/33549267
TB count            36102
TB avg target size  17 max=152 bytes
TB avg host size    668 bytes (expansion ratio: 37.7)
cross page TB count 35 (0%)
direct jump count   14863 (41%) (2 jumps=10899 30%)
TB hash buckets     4146/8192 (50.61% head buckets used)
TB hash occupancy   17.63% avg chain occ. Histogram: [0,10)%|█▁▅▁ ▂▁▁ ▁|[90,100]%
TB hash avg chain   1.008 buckets. Histogram: 1|█▁|2

Statistics:
TB flush count      13
TB invalidate count 324441
TLB flush count     484
JIT cycles          4737605000 (1.974 s at 2.4 GHz)
translated TBs      337178 (aborted=0 0.0%)
avg ops/TB          93.9 max=362
deleted ops/TB      2.09
avg temps/TB        41.03 max=46
avg host code/TB    1158.3
avg search data/TB  43.2
ops                 31648062
in bytes            10142744
out bytes           390555017
search bytes        14581774
cycles/op           149.7
cycles/in byte      467.1
cycles/out byte     12.1
cycles/search byte     324.9
  gen_interm time   20.5%
  gen_code time     79.5%
optim./code time    23.0%
liveness/code time  9.1%
cpu_restore count   106
  avg cycles        160.4
```

What's particularly interesting is the cycles/op number: 149.7. That's the total cycles spent doing compiling (measured by rdtsc) divided by the number of qemu IR operations generated. I doubt binaryen would be that fast. 20% of the time here is spent doing compiling, so a slowdown in that area would be noticeable.

@kripken

**Comment by kripken** [comment]
Binaryen would add to that, yeah. But, being realistic, Binaryen is going to be faster than the WebAssembly VM compiling it to machine code afterwards anyhow. WebAssembly is an intermediate format, so in an environment like qemu slower compilation is unavoidable.

In a workload dominated by compilation WebAssembly just might not be good enough, period. On the other hand, for code that is compiled once and run many times, increasing the compile time even significantly shouldn't be too bad - that's why I'm very interested in this. But, what's your use case?

**Comment by tbodt** [comment]
My usecase is having a nice development environment running on an iOS device. Qemu + WebAssembly seemed like the most promising way to make this fast.

**Comment by kripken** [comment]
I see. My guess is that could be fast enough (but we'd need to measure to be sure, of course).

Happy to help try this out - I can help writing out the Binaryen integration bits, if you set up a repo. I just don't know where to get started on the qemu side.

Btw, another option here is to interpret the wasm (using the Binaryen interpreter, or another), to avoid compilation entirely. Then compilation can be done after the function is used enough times, as a JIT optimization, etc. I'm optimistic we can find ways to make this work efficiently.

**Comment by davidar** [comment]
Did this go anywhere? I'd also be interested in seeing this happen

**Comment by tbodt** [comment]
@davidar No, I never found the time to work on it. It's still on my list of unfinished projects that I might work on later though.

**Comment by davidar** [comment]
@atrosinenko You seem to have quite some experience [building QEMU with Emscripten](https://github.com/atrosinenko/qemujs), I'm curious if you have any thoughts on this issue?

**Comment by atrosinenko** [comment]
@davidar Now I'm planning to rewrite my port of QEMU on top of v3.0 implementing WASM JIT instead of Asm.js one that was implemented more as proof-of-concept (and fall back to compiling plain TCI in Asm.js) -- see `qemujs-v2` branch. But I fear Binaryen's Apache 2.0 license is not compatible with QEMU that is AFAIK GPLv2 as a whole (different parts have different license options sometimes).

For Asm.js I was compiling basic block after N executions (hooked `qemu_tb_exec` -- you can see diff of v2.4.1...emscripten branches) and hope it will be much faster when eliminating `machine code -> TCI bytecode -> asm.js -> parse JS -> ... -> machine code` flow.

**Comment by kripken** [comment]
Regarding the license, that is something we should fix if, if it's a problem.

@jfbastien, I know the `WebAssembly/` projects, like LLVM, were thinking about doing Apache 2 plus exceptions. Looks like those exceptions might help here (QEMU is GPL2). Do you know what the status is there?

**Comment by atrosinenko** [comment]
> Regarding the license, that is something we should fix if, if it's a problem.

But isn't changing a license a big problem when there are already 64 contributors (or do you have copyright transfer)? (I'm not a lawyer, maybe Apache 2.0 is not a problem...)

Do you mean **upstreaming** Binaryen support for QEMU? If so, it may be worth reading QEMU contributor guidelines beforehand: [Contribute/SubmitAPatch](https://wiki.qemu.org/Contribute/SubmitAPatch), to not fix formal requirements afterwards. But not sure whether they would accept such a port. For example, AFAIK now multithreading in browsers is disabled due to security reasons, so I have some kludges to make QEMU single-threaded at the cost of stability. On the other hand, if the original issue was about native QEMU linked to system JS/WASM-engine, then only CPU thread would be "single-threaded", so why not... Anyways, it would be great to upstream as much as possible.

Meanwhile, is there some possibility in WASM to compile one basic block at a time to WASM modules and then, when QEMU decides to directly link them, somehow link these modules, so they can pass execution from one piece of AOT compiled code to another one efficiently?

**Comment by tbodt** [comment]
@atrosinenko Multithreading was recently re-enabled (in Chrome at least) because other spectre mitigations were put in place.

**Comment by tbodt** [comment]
As for compiling one basic block at a time, you can either:

 - Put all the basic blocks in a function, run binaryen's relooper on the function, and reload the module
 - Wait for tailcall support, make each basic block its own function, and reload the entire module with all the basic blocks

Both solutions require the browser to throw out all of its existing compiled machine code and start over. It seems like WASM wasn't designed with this use case in mind.

**Comment by jfbastien** [comment]
@kripken http://llvm.org/foundation/relicensing/

**Comment by binji** [comment]
@tbodt you can also compile the function in its own module, export it, and put it in a table. Then the original module can use `call_indirect` to call this function.

**Comment by kripken** [comment]
@jfbastien Thanks! And to be sure, the wasm projects are going to be licensed in the exact same way as LLVM?

@atrosinenko 

> But isn't changing a license a big problem when there are already 64 contributors?

Well, in practice probably 99% of the code in Binaryen was written by less than 10 people. And I'm pretty sure all of them are very excited about enabling use cases like Qemu (I definitely am), so I think we could dual-license it (say, to add MIT) or some other solution. But it sounds like the LLVM license change will apply here too, in a matter of months, so we may not even need to do that.

Overall, I encourage you to experiment on this stuff :) We'll make the licensing stuff work out.

**Comment by jfbastien** [comment]
> @jfbastien Thanks! And to be sure, the wasm projects are going to be licensed in the exact same way as LLVM?

That was discussed a long time ago. Probably worth bringing up again at the CG meeting. @binji ?

**Comment by binji** [comment]
@jfbastien I'm definitely not up-to-speed on WebAssembly licensing stuff -- but happy to add it to the agenda if someone can drive the conversation. :-)

**Comment by kripken** [comment]
I'm happy to drive the conversation. I thought I remembered we had a plan here and/or someone that was focused on this? But I guess we can discuss that in the CG meeting :)

**Comment by binji** [comment]
@kripken Thanks, I've added an agenda item here: https://github.com/WebAssembly/meetings/pull/296. We have a pretty full meeting, so it may end up pushed to the next one.

**Comment by jfbastien** [comment]
@kripken can you dig up the prior discussions for this?

**Comment by binji** [comment]
It came up in the CG here as well: https://github.com/WebAssembly/meetings/blob/master/2018/CG-01-26.md#contribution-policies-and-repos-for-tools-targeting-webassembly

With slides: https://docs.google.com/presentation/d/1sgxnT7pD5bTfmPv6MLnFlzIcszUDyTc9SKiSV6LZAJk/edit?usp=sharing

Doesn't seem like there was a follow-up.

@dschuff ?

**Comment by dschuff** [comment]
OK, well there are 2 issues in play here. One is the license and the other is whether you need to be in the CG in order to contribute. We only discussed the latter, nobody proposed changing the license.

Brad and I talked with a lawyer about the CG membership requirement, but to be honest they weren't super helpful, in the sense that there wasn't really any specific guidance, or particular legal thing that we learned. Mostly it's just the idea that there might be an increase in risk that something gets into the spec via discussion on the tools or some forum that's not covered by the CG's legal agreements, and we have to decide if we want to accept that. My own thought is that 1) I think the risk in general is quite small, and historically we've been diligent about pushing discussion of wasm itself and the spec into the design repo or other appropriate places, and 2) this really just makes things consistent, i.e. LLVM and the browsers are written to the spec, and they are not covered by the CG, so changing Binaryen and Wabt to match them would not really change anything substantive. So I'm still fine with making that change in principle. We didn't discuss how to actually implement that.

But I'd also guess that changing the license would be a bigger deal.

**Comment by jfbastien** [comment]
Latest update on LLVM relicensing: http://lists.llvm.org/pipermail/llvm-foundation/2018-July/000162.html
@dschuff I'd suggest you reach out to Danny (and say hi for me!). I had talked to him about it maybe ~3 years ago?

**Comment by atrosinenko** [comment]
@dschuff 

> But assuming we can work it all out with the Binaryen contributors, I'm also interested it knowing from the OP or other interested parties whether this would actually solve their problem.

As an interested party, I don't know right now whether Binaryen can help me :) , I was just asked what I think on using Binaryen as a QEMU backend and I answered on what I consider a blocker (maybe it isn't really). Now you say it can be fixed, then I need to re-evaluate.

Do I get it right, Binaryen can be used as a library constructing WASM binaries at run-time and even somehow optimize them? Meanwhile, is there some "official" WASM interpreter in C or C++? Another possibility may be to implement emitting WASM on my own, it would be some difficulty to master its binary format, but not sure whether I need much more than that from Binaryen.

**Comment by kripken** [comment]
> Do I get it right, Binaryen can be used as a library constructing WASM binaries at run-time and even somehow optimize them?

Yes, and that's one of the goals of Binaryen - to make it easy to write not just static compilers, but also JITs. I hope to see it used in places like QEMU and Mono, for example.

Specifically, Binaryen has a very simple C API for generating its IR, and then you can tell it to emit wasm from that. Optionally you can also run Binaryen's optimization passes first, which are designed to be fast enough to run in a JIT.

**Comment by atrosinenko** [comment]
Sounds very promising!

**Comment by binji** [comment]
> Meanwhile, is there some "official" WASM interpreter in C or C++?

The WebAssembly reference interpreter is written on OCaml: https://github.com/WebAssembly/spec/tree/master/interpreter

Binaryen and [wabt](https://github.com/WebAssembly/wabt) also have interpreters.

> Another possibility may be to implement emitting WASM on my own,

This is not as difficult as you might expect; WebAssembly is a relatively simple format. See, for example, [waforth](https://github.com/remko/waforth) which JITs Forth code.

**Comment by atrosinenko** [comment]
[Here](https://atrosinenko.github.io/qemujs-demo/new/shell.html) is a WIP example. Right now, the block layer does not work, since I had troubles with compiling with Emterpretify (hard to specify what to Emterpret and what not).

**Comment by kripken** [comment]
I get an out of memory crash in chrome, and on firefox I see an exception thrown, `exception thrown: CompileError: at offset 962: unrecognized opcode: c0 0` - is that expected?

**Comment by atrosinenko** [comment]
On Chrome it is expected and I don't know how to work around it generally, except for fixing Chrome. :)

In Firefox it should run Memtest (and then wait forever), OpenWRT expected to run for a while, then crash (everything tested without Network support).

**Comment by kripken** [comment]
I see, thanks. Ok, let's file a bug on chrome for that, yeah. First thing, though, is that an optimized build? I see the binaryen optimizer can shrink that file by 13%, which suggests maybe it's -O0?

**Comment by atrosinenko** [comment]
I fear, it is not because of the main WASM module, but because of ~1000th one -- when I debugged it the last time it successfully compiled 1000+ small WASM modules, then crashed, the same with Firefox, but with a much large module number. It would probably be quite easy to compile not everything if I manage to set up interpretation in Bynarien. 

Compiling selected TBs only after some execution count would probably be much faster, as well.

**Comment by kripken** [comment]
I see, thanks. Ok, I can open a bug with that - is it a stable URL?

**Comment by atrosinenko** [comment]
@kripken I copied it to separate directory: https://atrosinenko.github.io/qemujs-demo/chrome-bug/shell.html -- it will probably be deleted sometime, but can be considered as a stable URL for the time of fixing the bug.

**Comment by kripken** [comment]
Thanks! I filed https://bugs.chromium.org/p/v8/issues/detail?id=8907

**Comment by atrosinenko** [comment]
Thanks! Now I work on using Binaryen interpreter for first N executions, it even run Memtest in Chrome, but now it leaks (in a C meaning of leak) even in Firefox, so I try to run in natively but with Binaryen instead of TCI / native TCG. 

@tbodt Maybe it can somehow be used on iOS semi-natively this way (with JITting through the standard system JS engine), but I know almost nothing about iOS development.

**Comment by atrosinenko** [comment]
The demo was updated: now it does not translate everything, just TBs that are frequently executed. Meanwhile: how to handle Apache 2.0 vs GPLv2?

**Comment by kripken** [comment]
If the only problem with Apache is from Binaryen itself, then I'm pretty sure we can relicense it if we need to - not that old a project, not too many developers, and I doubt anyone would object. It would take some time and effort though.

However, the key issue might be whether the combined program makes [binaryen a derivative work of qemu](https://www.apache.org/licenses/GPL-compatibility.html). It might be worth asking a lawyer here, since it's not clear you are not really mixing Apache and GPL2 code - any new code in qemu to use Binaryen would be GPL2 I assume, and otherwise you are using Binaryen in an unmodified way through a pre-existing API. Anyhow, if it's hard to get a clear legal answer, or if the answer is negative, we can look into relicensing as mentioned above.

**Comment by vshymanskyy** [comment]
@atrosinenko You may want to look at Wasm3, which is a fast Wasm interpreter

**Comment by tlively** [comment]
Closing this because there doesn't seem to be anything actionable for Binaryen to do here. If the license issues are still relevant, we should discuss them in a new issue.
